File: Example-Mac-ObjC/ViewController.m
//
//  ViewController.m
//  Example-Mac-ObjC
//
//  Created by Darren Jones on 02/01/2020.
//  Copyright © 2020 Darren Jones. All rights reserved.
//

#import "ViewController.h"
#import <DJLogging/DJLogging.h>
#import "DJLogTypeUI.h"
#import "DJLogTypeComms.h"

@interface ViewController () <NSSharingServiceDelegate>

@end

@implementation ViewController

- (void)viewDidLoad
{
    // This method is called before AppDelegate's applicationDidFinishLaunching
    [LogManager setDebugLogsToScreen:YES];
    LogMethodCallWithType(DJLogTypeUI.shared)
    
    [super viewDidLoad];
    
    [self makeAWebRequest];
}

- (void)makeAWebRequest
{
    NSUUID *uuid = [NSUUID UUID];
    LogMethodCallWithUUIDAndType(uuid, DJLogTypeComms.shared)
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"https://venderbase.com"]];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:request
                                            completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        LogRequestResponseWithUUIDAndType(uuid, DJLogTypeComms.shared)
    }];
    [task resume];
}

- (IBAction)emailSupportButtonPressed:(id)sender
{
    LogMethodCall
    
    // Save HTML as temp file
    NSData *htmlData = [LogManager htmlData];
    NSURL *tempFile = [NSURL fileURLWithPath:[NSTemporaryDirectory() stringByAppendingFormat:@"logs.html"]];
    NSLog(@"%@", tempFile);
    
    // Check if file exists
    if ([[NSFileManager defaultManager] fileExistsAtPath:tempFile.path] == YES)
    {
        [[NSFileManager defaultManager] removeItemAtPath:tempFile.path error:nil];
    }
    
    BOOL success = [htmlData writeToFile:tempFile.path atomically:YES];
    if (success)
    {
        // Open email with attachment
        [self emailSupportWithAttachment:tempFile];
    }
    else
    {
        [self displayModalAlertWithTitle:@"Error"
                                 message:@"Failed to attach logs to email"
                             buttonTitle:@"OK"
                              alertStyle:NSAlertStyleInformational];
        
        [self emailSupportWithAttachment:nil];
    }
}

#pragma mark - NSSharingService - Email

- (void)emailSupportWithAttachment:(NSURL *)attachment
{
    NSString *to = @"add.support.email@here.com";
    NSString *appName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleName"];
    NSString *subject = [NSString stringWithFormat:@"%@ Support Request (MacOS)", appName];
    NSString *body = @"\n\n\n\n";
    
    NSSharingService *sharingService = [NSSharingService sharingServiceNamed:NSSharingServiceNameComposeEmail];
    [sharingService setRecipients:@[to]];
    [sharingService setSubject:subject];
    sharingService.delegate = self;
    
    if (attachment)
    {
        [sharingService performWithItems:@[body, attachment]];
    }
    else
    {
        [sharingService performWithItems:@[body]];
    }
}

- (void)sharingService:(NSSharingService *)sharingService didFailToShareItems:(NSArray *)items error:(NSError *)error
{
    [self displayModalAlertWithTitle:@"Error"
                             message:@"Failed to open mail composer"
                         buttonTitle:@"OK"
                          alertStyle:NSAlertStyleInformational];
}

#pragma mark - NSAlert

- (void)displayModalAlertWithTitle:(NSString *)title message:(NSString *)message buttonTitle:(NSString *)buttonTitle alertStyle:(NSAlertStyle)alertStyle
{
    dispatch_async(dispatch_get_main_queue(), ^{
        NSAlert *alert = [[NSAlert alloc] init];
        [alert setMessageText:title];
        [alert setInformativeText:message];
        [alert addButtonWithTitle:buttonTitle];
        [alert setAlertStyle:alertStyle];
        [alert runModal];
    });
}

@end

File: Example-Mac-Swift/ViewController.swift
//
//  ViewController.swift
//  Example-Mac-Swift
//
//  Created by Darren Jones on 02/01/2020.
//  Copyright © 2020 Darren Jones. All rights reserved.
//

import Cocoa
import DJLogging

class ViewController: NSViewController, NSSharingServiceDelegate {

    override func viewDidLoad() {
        
        LogManager.setDebugLogsToScreen(true)
        
        LogMethodCall(type: .ui)
        
        super.viewDidLoad()
        
        makeAWebRequest(url: URL(string: "https://venderbase.com")!)
        makeAWebRequest(url: URL(string: "https://somethingnotvalidxxx.com")!)
    }
    
    func makeAWebRequest(url: URL) {
        let uuid = UUID()
        LogMethodCall(uuid, type: .comms)
        
        let request = URLRequest(url: url)
        let session = URLSession(configuration: .default)
        let task = session.dataTask(with: request) { (data, response, error) in
            
            LogRequestResponse(uuid: uuid, response: response, data: data, error: error, type: .comms)
        }
        task.resume()
    }

    @IBAction func emailSupportButtonPressed(_ sender: Any) {
        
        LogMethodCall(type: .ui)
        
        // Save HTML as temp file
        let htmlData = LogManager.htmlData() ?? Data()
        let tempFile = NSTemporaryDirectory().appending("logs.html")
        print(tempFile)
        
        do {
            let fileManager = FileManager.default

            // Check if file exists
            if fileManager.fileExists(atPath: tempFile)
            {
                // Delete file
                try fileManager.removeItem(atPath: tempFile)
            } else {
                print("File does not exist")
            }
            
            let tempFileURL = URL(fileURLWithPath: tempFile)
            try htmlData.write(to: tempFileURL)
            
            emailSupportWithAttachment(tempFileURL)

        }
        catch let error as NSError {
            print("An error took place: \(error)")
            
            displayModalAlertWithTitle("Error", message: "Failed to attach logs to email", buttonTitle: "OK", alertStyle: .informational)
            
            emailSupportWithAttachment(nil)
        }
    }
    
    // MARK: - NSSharingService - Email
    
    func emailSupportWithAttachment(_ attachment:URL?) {
        
        let to = "add.support.email@here.com"
        let appName = Bundle.main.object(forInfoDictionaryKey: "CFBundleName")
        let subject = "\(appName ?? "") Support Request (MacOS)"
        let body = "\n\n\n\n"
        
        let sharingService = NSSharingService.init(named: .composeEmail)
        sharingService?.recipients = [to]
        sharingService?.subject = subject
        sharingService?.delegate = self
        
        if attachment != nil
        {
            sharingService?.perform(withItems: [body, attachment!])
        }
        else
        {
            sharingService?.perform(withItems: [body])
        }
    }
    
    func sharingService(_ sharingService: NSSharingService, didFailToShareItems items: [Any], error: Error) {
        
        displayModalAlertWithTitle("Error", message: "Failed to open mail composer", buttonTitle: "OK", alertStyle: .informational)
    }
     
     // MARK: - NSAlert

    func displayModalAlertWithTitle(_ title:String, message:String, buttonTitle:String, alertStyle:NSAlert.Style) {
        
        DispatchQueue.main.async {
            let alert = NSAlert()
            alert.messageText = title
            alert.informativeText = message
            alert.addButton(withTitle: buttonTitle)
            alert.alertStyle = alertStyle
            alert.runModal()
        }
    }
    
}

